Pl add custom approvals topic doc 2594 (#675)
harness/developer-hub: docs/platform/9_Approvals/custom-approvals.md

---
title: Adding Custom Approval Stages and Steps
description: Use a Custom Approval stage and Custom Approval step to approve or reject a Pipeline or stage at any point in its execution
sidebar_position: 40
helpdocs_topic_id: 9ulc3vrpf2
helpdocs_category_id: 2d7y1cr09y
helpdocs_is_private: false
helpdocs_is_published: true
---

> Currently, this feature is behind the feature flag `NG_CUSTOM_APPROVAL`. Contact [Harness Support](mailto:support@harness.io) to enable the feature.Harness has first class support for [Jira](adding-jira-approval-stages.md) and [ServiceNow](service-now-approvals.md) approval stages and steps, and Harness has a built-in [Manual](adding-harness-approval-stages.md) approval stages and steps.
> 
> In some cases, you might have a custom shell script ticketing system or prefer to script your approvals. You can use the Custom Approval stage and step with these methods.
> 
> A Custom Approval stage and step can approve or reject a Pipeline or stage at any point in their execution. You add a script to a Custom Approval step and use the script results as approval and rejection criteria.

Reviewer: ```suggestion
Currently, this feature is behind the feature flag `NG_CUSTOM_APPROVAL`. Contact [Harness Support](mailto:support@harness.io) to enable the feature.

Custom approval stages and steps add control gates to your pipelines by allowing you to approve or reject a pipeline or stage at any point during build execution. When you add a Custom Approval step, you add a script to the step, and then use the script results as approval or rejection criteria.

You might use a Custom Approval stage and step, for example, if you have a custom shell script ticketing system or if you prefer to script your approvals.

Harness has first class support for [Jira](adding-jira-approval-stages.md) and [ServiceNow](service-now-approvals.md) approval stages and steps, and Harness has a built-in [Manual](adding-harness-approval-stages.md) approval stages and steps.

```

Missing line break after feature flag disclaimer.
Reorganized ideas to describe what a Custom Approval does before describing why you would use it and what support Harness has.

This topic describes how to add Custom Approval stages and steps.

### Before You Begin

* [Learn Harness' Key Concepts](../../getting-started/learn-harness-key-concepts.md)

> * [Add a Stage](../8_Pipelines/add-a-stage.md)

Reviewer: This process does not really have a true prerequisite (except perhaps having a pipeline to add the stage to). I would consider adding these links to "See also", at the end, or combine them as a single sentence with the introduction, rather than using a separate heading.

> ### Requirements and Limitations

Reviewer: ```suggestion
## Requirements and limitations
```
Capitalization.

Heading levels need to start with level 2. The page title is level 1, so starting with level 3 skips a level.

> * A Custom Approval stage can be added to any Pipeline created in any module, such as in Continuous Delivery (CD), Continuous Integration (CI), or Feature Flags.

Reviewer: ```suggestion
* A Custom Approval stage can be added to any pipeline created in any module, such as in Continuous Delivery (CD), Continuous Integration (CI), or Feature Flags.
```

* A Custom Approval step can be added to a Custom Approval stage, CD Build stage, or Feature Flags stage.

> * A Custom Approval step is not available in the the CI stage currently.

Reviewer: ```suggestion
* A Custom Approval step is not available in the CI stage currently.
```

repeated word

> * The Custom Approval step must have at least one Approval criteria. Rejection criteria is optional.

Reviewer: ```suggestion
* A Custom Approval step must have at least one Approval criteria. Rejection criteria is optional.
```

Parallel construction

> ### Review: Custom Approval Stages and Steps

Reviewer: ```suggestion
## Review: Custom Approval stages and steps
```

Sentence case for capitalization in headings.

TBH I would probably move all or part of this section into the intro. Maybe the "Criteria can be the follow:" list could go closer to the step where the user is configuring the script or criteria.

Criteria can be the following:

* Evaluating a script output.
	+ For example, you set a value for a variable in your script, output that variable from the script, and then use that variable in the approval/rejection criteria of the Custom Approval step.
	+ When an evaluation results in a Boolean true, the approval criteria is met.
* Evaluating a JEXL Expression.
	+ Harness supports JEXL expressions and you can use these with script outputs or Harness variables to create expressions.
	+ If the resolved JEXL expression is a Boolean true, the approval criteria is met.
	+ For example, you could evaluate the output from a previous [Shell Script](../../continuous-delivery/cd-execution/cd-general-steps/using-shell-scripts.md) step named **test** by referencing its output in a JEXL expression and comparing it to the value `Approved`: `<+execution.steps.test.output.outputVariables.status>=="Approved"`.
	+ See [Built-in and Custom Harness Variables Reference](../12_Variables-and-Expressions/harness-variables.md) and [Extracting Characters from Harness Variable Expressions](../12_Variables-and-Expressions/extracting-characters-from-harness-variable-expressions.md).

> ### Step 1: Add a Custom Approval Stage

Reviewer: ```suggestion
## Add a Custom Approval stage
```

Use numbering for the steps in these sections.

The section title is "add a custom approval stage" but the steps talk about adding a custom approval to an existing stage, and there is also a lot of info about the step settings here. A better title might be something like "Add a Custom Approval to a pipeline"?

> A Custom Approval stage contains a Custom Approval step, but you can also add a Custom Approval step to other Approval stages, as well as CD Build and Feature Flags stages.
> 
> Let's add a Custom Approval stage and see the Custom Approval step added automatically.

Reviewer: ```suggestion
Custom Approval stages automatically contain Custom Approval steps, and you can add Custom Approval steps to other Approval stages, CD Build stages, and Feature Flags stages.
```

IIRC the style guide stated not to use phrasing like "Let's do...".

In a Pipeline, click **Add Stage.**

In **Select Stage Type**, click **Approval**.

![](./static/custom-approvals-00.png)

In **About your Stage**, enter a name for this stage, select **Custom**, and click **Set Up Stage**. The Approval Stage is set up.

![](./static/custom-approvals-01.png)

Click the **Custom Approval** step to set up your **Step Parameters**.

![](./static/custom-approvals-02.png)

> If you were adding a Custom Approval step to a different type of stage, you would simply click **Add Step** in that stage's **Execution** and select **Custom Approval**.### Step 2: Set up a Custom Approval Step

Reviewer: You could use tabs to show two sets of steps like:

- Add Custom Approval stage
- Add Custom Approval step to existing stage

![](./static/custom-approvals-03.png)

The format to reference the output variable is `<+steps.[step_id].output.outputVariables.[output_variable_name]>`.

Here's an example showing how the **Script Output Variables** references the exported variable **test**, and how you reference the output variable Name **var1** in a subsequent Shell Script step to get that value:

![](./static/custom-approvals-04.png)

To find the expression to reference your output variables, execute the Pipeline, locate the step, and click its **Output** tab. You can copy the Output variable there.

![](./static/custom-approvals-05.png)

#### Using Variables in Approval or Rejection Criteria

You can use script variables in your approval or rejection criteria by use them in **Script Output Variables**.

First, you create the variable in **Script**, then create a variable in **Script Output Variables** that references that variable, and then use the **Script Output Variables** variable and the **Script** variable's value in the **Approval Criteria**:

![](./static/custom-approvals-06.png)

> ### Step 3: Set Approval Criteria

Reviewer: ```suggestion
## Set Approval Criteria
```

I'm not sure whether approval criteria should be capitalized or not. Technically it's capitalized in the UI.

The **Approval Criteria** in the step determines if the Pipeline or stage is approved.

> **Approval Criteria** is mandatory, but you can also set **Rejection Criteria**.

Reviewer: ```suggestion
Custom Approval steps must have at least one **Approval Criteria** is mandatory. You can also specify **Rejection Criteria**.
```

![](./static/custom-approvals-07.png)

Whether the Pipeline/stage stops executing depends on the stage or step [Failure Strategy](../8_Pipelines/define-a-failure-strategy-on-stages-and-steps.md).You can specify criteria using **Conditions** and/or **JEXL Expression**. If you use them in combination, they both must evaluate to a Boolean `True` for the step to be approved.

In **Conditions**, you can define approval criteria using outputs from the step script, [Harness expressions](../12_Variables-and-Expressions/harness-variables.md), or your custom ticketing system.

If you use Harness expressions, ensure that the expression can be resolved at runtime.

If you use custom variable expressions that reference information from subsequent steps, ensure that those custom variable expressions will be resolved.

In **JEXL Expression**, you can use [JEXL expressions](https://commons.apache.org/proper/commons-jexl/reference/syntax.html). You can use a JEXL expression if the field is set to **Fixed value** or **Expression**.

> ### Option: Script Input Variables

Reviewer: All of these "option" subsections could be accordions or tabs. use of "Option" makes it seem like the user doesn't have to do any of them. While that may be true, there is no subheading marked at required (or at least not "Option") so it seems like it's all optional.

### Option: Set Rejection Criteria

In **Optional Configuration**, in **Rejection Criteria**, you can define criteria for rejecting the approval.

If you add rejection criteria it is used in addition to the settings in **Approval Criteria**. Both **Rejection Criteria** and **Approval Criteria** must evaluate to a Boolean true for the step to be successful.

### Option: Harness Expressions in Variables

You can use [Harness variable expressions](../12_Variables-and-Expressions/harness-variables.md) in your scripts and in the **Script Input Variables** and **Script Output Variables**.

For **Script Input Variables** and **Script Output Variables**, you simply select **Expression**, and then paste in the Harness variable expression.

### Option: Advanced Settings

In **Advanced**, you can use the following options:

* [Delegate Selector](../2_Delegates/manage-delegates/select-delegates-with-selectors.md#option-select-a-delegate-for-a-step-using-tags)
* [Step Skip Condition Settings](../8_Pipelines/w_pipeline-steps-reference/step-skip-condition-settings.md)
* [Step Failure Strategy Settings](../8_Pipelines/w_pipeline-steps-reference/step-failure-strategy-settings.md)

> ### Step 3: Apply and Test

Reviewer: ```suggestion
## Apply and Test
```

I think these steps could be combined with the above section --  such as "Configure and test criteria".

Click **Apply Changes**. The Custom Approval step is added to the stage.

Run the Pipeline.

> When the Custom Approval step is reached, you can see its approval and rejection criteria.

Reviewer: ```suggestion
When the Custom Approval step is reached, you can see its approval and rejection criteria.
```

Does this part of the pipeline shows the criteria the user provided, or the outcome of the evaluation?

> ### YAML Example

Reviewer: ```suggestion
## YAML example
```

                          - step:  
                                type: ShellScript  
                                name: vars  
                                identifier: vars  
                                spec:  
                                    shell: Bash  
                                    onDelegate: true  
                                    source:  
                                        type: Inline  
                                        spec:  
                                            script: |-  
                                                echo "output var: "<+steps.custapprove.output.outputVariables.var1>  
  
                                                test=<+steps.custapprove.output.outputVariables.var1>  
                                    environmentVariables: []  
                                    outputVariables:  
                                        - name: foo  
                                          type: String  
                                          value: test  
                                    executionTarget: {}  
                                timeout: 10m  
              tags: {}  

```

> ### Notes

Reviewer: ```suggestion
## Notes
```

I would move these subsections into more relevant areas of the topic. Most of these are relevant to writing the script, so they would be more useful there. The "Published variables not available" might belong after the "Run the pipeline" test, since I am assuming they would first see that error during the test run.
